### ISSUE TITLE
Physics calculations in constant time

### ISSUE THREAD

**Issue by altf4** [altf4/SmashBot#85]
I have some really hacked up functions that do physics calculations for SmashBot. They **work** but are very bad and are starting to have issues. 

In particular, they essentially do multiplication by repeated addition:

https://github.com/altf4/SmashBot/blob/master/Util/GameState.cpp#L1609

This has caused SmashBot to sometimes deadlock, being stuck in that loop. Even though I added a kill switch for a max iteration count... it's still just not a good algorithm. Needs improving to be constant time.

Other such functions:

https://github.com/altf4/SmashBot/blob/master/Util/GameState.cpp#L1175
https://github.com/altf4/SmashBot/blob/master/Util/GameState.cpp#L1472

**Comment by spxtr** [comment]
[Kinematic equations](https://en.wikipedia.org/wiki/Equations_of_motion#Uniform_acceleration) will help for the first.

**Comment by geranim0** [comment]
you can precompute this

**Comment by altf4** [comment]
@geranim0 The first function can, but not the other two. The other functions allow the caller to specify a frame count. Essentially asking "How far will this character slide after exactly X frames". It would be impractical to precompute that, especially since there ought to be a constant time solution.

**Comment by geranim0** [comment]
Well you could always precompute the slide distance as a function some sliding coefficients, initspeeds and frames, and linearly interpolate if the input is between 2 precomputed values

**Comment by altf4** [comment]
That sounds hard. The math is already correct, I just need it to not be a linear time algorithm. I think @spxtr is right.

**Comment by Eppie** [comment]
Take a look at this modification (using `initSpeed * initSpeed` just to save the call to `std::pow()`:
```c++
double GameState::calculateDoubleJumpHeight(CHARACTER character, double initSpeed)
{
    return (initSpeed * initSpeed) / (2 * getGravity(character))
}
```

I compared the original function with this version, using Marth's gravity:
```
initSpeed: 0.1, original: 0.015, new: 0.05882352941176471
initSpeed: 0.2, original: 0.14500000000000002, new: 0.23529411764705885
initSpeed: 0.3, original: 0.38999999999999996, new: 0.5294117647058822
initSpeed: 0.4, original: 0.75, new: 0.9411764705882354
initSpeed: 0.5, original: 1.2249999999999999, new: 1.4705882352941175
initSpeed: 0.6, original: 1.8199999999999998, new: 2.117647058823529
initSpeed: 0.7, original: 2.5399999999999996, new: 2.88235294117647
initSpeed: 0.8, original: 3.3750000000000004, new: 3.7647058823529416
initSpeed: 0.9, original: 4.324999999999999, new: 4.764705882352941
initSpeed: 1.0, original: 5.390000000000001, new: 5.88235294117647
```
and with Fox's:
```
initSpeed: 0.1, original: 0, new: 0.02173913043478261
initSpeed: 0.2, original: 0, new: 0.08695652173913045
initSpeed: 0.3, original: 0.06999999999999998, new: 0.19565217391304346
initSpeed: 0.4, original: 0.17, new: 0.3478260869565218
initSpeed: 0.5, original: 0.31, new: 0.5434782608695652
initSpeed: 0.6, original: 0.51, new: 0.7826086956521738
initSpeed: 0.7, original: 0.7199999999999999, new: 1.0652173913043477
initSpeed: 0.8, original: 1.02, new: 1.3913043478260871
initSpeed: 0.9, original: 1.32, new: 1.7608695652173914
initSpeed: 1.0, original: 1.7000000000000002, new: 2.1739130434782608
```

I suppose the difference is because the original takes into account the fact that the game occurs in discrete time units (frames) whereas my version does not. I got the formula from [here](http://goo.gl/LewYjV).

**Comment by Eppie** [comment]
I have a partial solution for `double GameState::calculateSlideDistance(CHARACTER character, double initSpeed, int frames)`:

```c++
/* ... slideCoefficient calculation ... */
double result = frames * abs(initSpeed) - (slideCoefficient * (((frames * frames) + frames) / 2))
if(initSpeed < 0)
{
    return (-1.0) * result;
}
return result;
```
The `(((frames * frames) + frames) / 2)` part is just calculating the nth triangle number. I say "partial" solution because it doesn't work in all cases:
```
initSpeed: 0.25, frames: 1, original: 0.2, new: 0.2, diff = 0.0
initSpeed: 0.25, frames: 2, original: 0.35, new: 0.35, diff = 0.0
initSpeed: 0.25, frames: 3, original: 0.45, new: 0.45, diff = 0.0
initSpeed: 0.25, frames: 4, original: 0.5, new: 0.5, diff = 0.0
initSpeed: 0.25, frames: 5, original: 0.5, new: 0.5, diff = 0.0
initSpeed: 0.25, frames: 6, original: 0.5, new: 0.45, diff = -0.05
initSpeed: 0.25, frames: 7, original: 0.5, new: 0.35, diff = -0.15
initSpeed: 0.25, frames: 8, original: 0.5, new: 0.2, diff = -0.3
initSpeed: 0.25, frames: 9, original: 0.5, new: 0.0, diff = -0.5
initSpeed: 0.25, frames: 10, original: 0.5, new: -0.25, diff = -0.75
```
As you can see, it doesn't take into account the fact that the character needs to _stop_ sliding. Perhaps someone with more physics knowledge can improve upon this.

**Comment by altf4** [comment]
Closed for v2.0
